tools/crm_resource_print.c

```diff
@@ -306,6 +307,287 @@ property_text(pcmk__output_t *out, va_list args) {
     return pcmk_rc_no_output;
 }
 
+PCMK__OUTPUT_ARGS("resource-check", "pe_resource_t *", "char *", "char *")
+static int
+resource_check_default(pcmk__output_t *out, va_list args) {
+    pe_resource_t *rsc G_GNUC_UNUSED = va_arg(args, pe_resource_t *);
+    char *role_s G_GNUC_UNUSED = va_arg(args, char *);
+    char *managed G_GNUC_UNUSED = va_arg(args, char *);
+
+    pe_resource_t *parent = uber_parent(rsc);
+    int rc = pcmk_rc_no_output;
+    bool printed = false;
+
+    if (role_s) {
+        enum rsc_role_e role = text2role(role_s);
+
+        if (role == RSC_ROLE_UNKNOWN) {
+            // Treated as if unset
+
+        } else if (role == RSC_ROLE_STOPPED) {
+            if (!printed) {
+                printed = true;
+                out->begin_list(out, NULL, NULL, "Resource Checks");
+            }
+
+            out->list_item(out, "check", "Configuration specifies '%s' should remain stopped",
+                           parent->id);
+
+        } else if (pcmk_is_set(parent->flags, pe_rsc_promotable)
+                   && (role == RSC_ROLE_SLAVE)) {
+            if (!printed) {
+                printed = true;
+                out->begin_list(out, NULL, NULL, "Resource Checks");
+            }
+
+            out->list_item(out, "check", "Configuration specifies '%s' should not be promoted",
+                           parent->id);
+        }
+    }
+
+    if (managed && !crm_is_true(managed)) {
+        if (!printed) {
+            printed = true;
+            out->begin_list(out, NULL, NULL, "Resource Checks");
+        }
+
+        out->list_item(out, "check", "Configuration prevents cluster from stopping or starting unmanaged '%s'",
+                       parent->id);
+    }
+
+    if (rsc->lock_node) {
+        if (!printed) {
+            printed = true;
+            out->begin_list(out, NULL, NULL, "Resource Checks");
+        }
+
+        out->list_item(out, "check", "'%s' is locked to node %s due to shutdown",
+                       parent->id, rsc->lock_node->details->uname);
+    }
+
+    if (printed) {
+        out->end_list(out);
+        rc = pcmk_rc_ok;
+    }
```

Comment:
It's a bit of extra work, but I think it would lessen the chance of future divergence if we have a separate function for actually checking the various conditions, that just sets bits in a bitmask, and then the message functions check those bits. That might end up coming handy in other places anyway.

One issue is wanting to print lock_node. Maybe it could be a struct with a bitmask and lock_node.

That model would be harder to follow with resource-why.

I'm not set on that, we could leave that as an open question.

Reply:
Fixed.  I don't see a good way of reducing resource-why, though.  So much of that is iterating across lists hat it doesn't seem to lend itself well to the same model.